## Question 1

In [1]:
!pip install mlflow jupyter scikit-learn pandas seaborn hyperopt xgboost fastparquet boto3 optuna

You should consider upgrading via the '/Users/evanhofmeister/PycharmProjects/dezoomcamp-mlops-homework/venv/bin/python -m pip install --upgrade pip' command.


In [2]:
import mlflow
print(mlflow.__version__)

2.3.2


Answer: 2.3.2

## Question 2

In [3]:
import os
!mkdir os.getcwd()+'/data'

In [4]:
# Download data
!wget -P data https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-01.parquet
!wget -P data https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-02.parquet
!wget -P data https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-03.parquet

--2023-06-04 21:45:08--  https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-01.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 13.227.44.193, 13.227.44.50, 13.227.44.79, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|13.227.44.193|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1254291 (1.2M) [binary/octet-stream]
Saving to: 'data/green_tripdata_2022-01.parquet.4'

green_tripdata_2022 100%[===================>]   1.20M  6.85MB/s    in 0.2s    

2023-06-04 21:45:09 (6.85 MB/s) - 'data/green_tripdata_2022-01.parquet.4' saved [1254291/1254291]

--2023-06-04 21:45:09--  https://d37ci6vzurychx.cloudfront.net/trip-data/green_tripdata_2022-02.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 13.227.44.50, 13.227.44.79, 13.227.44.137, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|13.227.44.50|:443... connected.
HTTP re

![call preprocess script](preprocess_data.py)

In [5]:
!python preprocess_data.py --raw_data_path data --dest_path ./output

In [6]:
os.stat('output/dv.pkl').st_size

153660

Answer: 153660 bytes == 154 mega-bytes

## Question 3

![call train script](train.py)

In [7]:
!python train.py --data_path ./output

2023/06/04 21:45:26 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/evanhofmeister/PycharmProjects/dezoomcamp-mlops-homework/venv/lib/python3.9/site-packages/_distutils_hack/__init__.py:36: UserWarning: Setuptools is replacing distutils."


In [ ]:
# Start mlflow either from jupyter notebook or command line
!mlflow ui

[2023-06-04 21:33:17 -0500] [61591] [INFO] Starting gunicorn 20.1.0
[2023-06-04 21:33:17 -0500] [61591] [INFO] Listening at: http://127.0.0.1:5000 (61591)
[2023-06-04 21:33:17 -0500] [61591] [INFO] Using worker: sync
[2023-06-04 21:33:17 -0500] [61592] [INFO] Booting worker with pid: 61592
[2023-06-04 21:33:17 -0500] [61593] [INFO] Booting worker with pid: 61593
[2023-06-04 21:33:17 -0500] [61594] [INFO] Booting worker with pid: 61594
[2023-06-04 21:33:18 -0500] [61595] [INFO] Booting worker with pid: 61595


In [9]:
from mlflow.tracking import MlflowClient

client = MlflowClient()

# Retrieve the experiment ID from its name.
experiment_name = "random-forest"
experiment = client.get_experiment_by_name(experiment_name)
experiment_id = experiment.experiment_id

runs = client.search_runs(experiment_ids=[experiment_id])
for run in runs:
    params = client.get_run(run.info.run_id).data.params
    max_depth = params.get("max_depth")
    print(f"max_depth for run is {run.info.run_id}: {max_depth}")

max_depth for run is 0e7accf4ae67498fa2b841945be51cc7: 10
max_depth for run is 47d2d908478a402c89b393d7c42420e1: 10
max_depth for run is c0dc0c6a0991423480a9e93724daf785: 10
max_depth for run is 9acd543b186544eab564a7d0cedf7b8d: 10
max_depth for run is 567e1e8b071142a7ad249072d22ec096: 10


 Answer: 10

## Question 4

![call hpo script](hpo.py)

In [10]:
!python hpo.py --data_path ./output

2023/06/04 21:46:59 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.
[I 2023-06-04 21:46:59,287] A new study created in memory with name: no-name-2f7a17d4-1590-4df1-808a-a9d0050aaccd
[I 2023-06-04 21:47:00,069] Trial 0 finished with value: 2.451379690825458 and parameters: {'n_estimators': 25, 'max_depth': 20, 'min_samples_split': 8, 'min_samples_leaf': 3}. Best is trial 0 with value: 2.451379690825458.
[I 2023-06-04 21:47:00,155] Trial 1 finished with value: 2.4667366020368333 and parameters: {'n_estimators': 16, 'max_depth': 4, 'min_samples_split': 2, 'min_samples_leaf': 4}. Best is trial 0 with value: 2.451379690825458.
[I 2023-06-04 21:47:00,851] Trial 2 finished with value: 2.449827329704216 and parameters: {'n_estimators': 34, 'max_depth': 15, 'min_samples_split': 2, 'min_samples_leaf': 4}. Best is trial 2 with value: 2.449827329704216.
[I 2023-06-04 21:47:01,088] Trial 3 finished with value: 2.460983516558473 a

Answer: 2.45

## Question 5

![register_model script](register_model.py)

In [13]:
!python register_model.py --data_path ./output

Answer: 2.29

## Question 6

Answer: All of the above are true